index.js

```diff
@@ -1,5 +1,8 @@
 'use strict'
 
+const assert = require('assert')
```

Comment:
Thanks for being open to trying it. I'm a strong "Yay", explicitly only using `assert` for cases where it is a bug in our code.  Looking at both cases where I've used it here:

```
assert(part in completedFromPart, `'${part}' is in completedFromPart`)
```

I feel this is an easy one. This is an internal function only called inside that enclosing `makeIntakeRequest`. It is just a bug in that code if it is called with a value other than the designated three part names. Attempting to make `completePart()` somehow cope with an invalid `part` value and stumble on with the intake request would make the code more complex and would tend to hide the programmer error in dev and testing.


```
    assert(client._encodedMetadata, 'client._encodedMetadata is set')
```

This is the one we discussed last time. Basically I make the same arguments: The code is made more complex attempting to cope with `client._encodedMetadata` not being set. While the code flow is less obvious compared to the `assert` above, there is no valid usage of the Client that allows `makeIntakeRequest` to be called before `client._encodedMetadata` is set. At that point, we, the developers don't understand the program state and can't make any guarantees about behaviour.

Another comment on the "code is made more complex": I was looking at completing the work for adding a filter on metadata (https://github.com/elastic/apm-agent-nodejs/issues/1916) on Monday. I decided to wait for this work to go in, because the current coping code here is a *4th* place where `client._encode(toEncode, Client.encoding.METADATA)` is called -- and where a new metadata filter should be applied.

In general, in my view, the important distinction is operational errors vs. programmer errors (as covered by https://www.joyent.com/node-js/production/design/errors).  These two cases cover what would be programmer errors. I realize the advice there:

> The best way to recover from programmer errors is to crash immediately.

is possibly controversial.  I understand APM Agent authors need to be very concerned about adding possible crashes in the user's app. However, I'd advocating for using asserts only to check internal invariants in the agent code to simplify and to makes subtle bugs obvious ones.

Reply:
Sounds like a plan -- let's move forward with assertions for _programmer errors_ and not operational errors.  We can revisit the descision if/when we discover that the real world find ways to trigger our assertions.